### PR TITLE
Change target-typed-conditional to LDM-decided semantics

### PR DESCRIPTION
**Pull request by gafter**
- Use the same semantics for target-typed conditionals regardless of language version
  However, if the language version isn't sufficient, it is an error to *use* that conversion
- Update the semantics of the target-typed-conditional conversion per LDM
  See https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-conditional-expression.md (bottom)
- We still need to test against a trial VS insertion

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this change of the target-typed-conditional semantics?

**Comment by gafter** [comment]
This fails to build bootstrap:
```none
src\Compilers\Core\Portable\CvtRes.cs(566,36): error CS8652: The feature 'target-typed conditional expression' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
```

The problematic line is this
```csharp
resWriter.Write((WORD)((isDll) ? 0x0002 : 0x0001));
```

which, simplified, looks like this

```csharp
M((short)(someBool ? 1 : 2));
```

In C# 8.0, this was an explicit conversion. But due to the target-typed conversion, this is an implicit (target-typed conditional) conversion under the new rules. Since we always prefer an implicit conversion to an explicit one, we attempt to use the new conversion and fail due to the language version.

**Comment by RikkiGibson** [comment]
It might be good to do a test insertion to VS before merging this feature to master to see if VS is also affected by the breaking change.

**Comment by gafter** [comment]
@RikkiGibson You appear to have read the text in this PR's original comment.

**Comment by RikkiGibson** [comment]
Got it, thanks!

**Comment by gafter** [comment]
May I please have a couple of reviews of this revision of the target-typed-conditional feature to LDM-decided semantics?

**Comment by gafter** [comment]
May I please have a couple of reviews of this revision of the target-typed-conditional feature to LDM-decided semantics?

**Comment by gafter** [comment]
May I please have a couple of reviews of this revision of the target-typed-conditional feature to LDM-decided semantics?

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 1]
```diff
@@ -1537,16 +1537,17 @@ class C
 {
     void M1()
     {
-        object obj = true ? 0 : M2();
+        var obj = true ? 0 : M2();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:1540]: Why was this changed to 'var'?

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:1540]: Because this test would not test what it was intended to test if there is a target type of `object`. The test would produce no diagnostics under the new rules, but the point of the test is to demonstrate this diagnostic is produced.

---
In reply to: [441790482](https://github.com/dotnet/roslyn/pull/44505#discussion_r441790482) [](ancestors = 441790482)

**Comment by gafter** [comment]
Agreed (as pointed out in the OP).

---
In reply to: [634300288](https://github.com/dotnet/roslyn/pull/44505#issuecomment-634300288) [](ancestors = 634300288)
